cvat/apps/events/handlers.py

```diff
@@ -594,3 +588,96 @@ def handle_viewset_exception(exc, context):
     )
 
     return response
+
+def handle_client_events_push(request, data: dict):
+    TIME_THRESHOLD = datetime.timedelta(seconds=100)
+    WORKING_TIME_SCOPE = 'send:working_time'
+    WORKING_TIME_RESOLUTION = datetime.timedelta(milliseconds=1)
+    COLLAPSED_EVENT_SCOPES = frozenset(("change:frame",))
+    org = request.iam_context["organization"]
+
+    def read_ids(event: dict) -> tuple[int | None, int | None, int | None]:
+        return event.get("job_id"), event.get("task_id"), event.get("project_id")
+
+    def get_end_timestamp(event: dict) -> datetime.datetime:
+        if event["scope"] in COLLAPSED_EVENT_SCOPES:
+            return event["timestamp"] + datetime.timedelta(milliseconds=event["duration"])
+        return event["timestamp"]
+
+    def generate_wt_event(
+        ids: tuple[int | None, int | None, int | None],
+        wt: datetime.timedelta,
+        common: dict
+    ) -> EventSerializer | None:
+        job_id, task_id, project_id = ids
+        if wt.total_seconds():
+            value = wt // WORKING_TIME_RESOLUTION
+            event = EventSerializer(data={
```

Comment:
Only for extra validation
During development this validation helped me, and I believe it is a good idea to validate any data we are going to save in the database

Reply:
I'm not sure I agree. Validation is for transforming data from untrusted (i.e. coming from the client) to trusted. If the data is coming from the server, like here, then it is already trusted. Furthermore, using `EventSerializer` specifically is problematic for a few reasons:

* `Serializer.is_valid` raises a `ValidationError`, which is transformed by DRF to `400 Bad Request`, i.e. a client-side problem. However, if the data here is invalid, then the problem is on the server.

* `EventSerializer` is meant to validate events coming from the client, and we may not necessarily want the same requirements for events originating from the server.

Reply:
I'd also add that if you want to do any sanity checking, then the right place to do it would be in `record_server_event`, so that the checks are shared between all events.

Reply:
Alright, makes sence